### DISCUSSION TITLE
Edge IDs

### DISCUSSION THREAD

**Discussion by hchiba1** [pg-format/pg-formatter#34] [Ideas]
Some DB management systems may utilize edge IDs.

For such cases, we can consider adding an optional column at the beginning of `Edge`:
`101 a -> b`

**Comment by hchiba1** [comment]
**A confusing case:**

`1 -> -> a : b`
* edge from `1` to `->` with property `a : b`
* edge with **ID** `1` from `->` to `a` with label `:b`

**Reply by nichtich** [reply]
This would only be a problem if direction is not separated with whitespace and for undirected edges, as `<` and `>` are not allowed in unquoted identifiers. The string `--` could be excluded as well. Relate issue also https://github.com/orgs/pg-format/discussions/29.

In any case direction (`--`, `->`, `<-`) must never the confused with unquoted identifier.

**Reply by hchiba1** [reply]
I realized the following will work for excluding Direction from IDs:
```js
ID = !Direction s:StringNonEmpty
{
  return s;
}
```

**Comment by hchiba1** [comment]
In PG-JSONL, the solution would be straightforward, i.e., simply adding an `"id":` to each edge.

**Comment by hchiba1** [comment]
Would it be better not to support in version 1.0?

**Reply by nichtich** [reply]
For a general graph exchange format I'd like to have all of (in this order)

- edge ids
- graph attributes
- subgraphs/graph hierarchies
- hyper-edges

All are relatively easy in PG-JSON(L) but support in less so in PG format so these features could be postponded.

Some ideas on edge id syntax in PG format (edge with id `x` from `a` to `b`):

~~~
x a -> b
x: a -> b
a -> b ^x
^x a -> b
~~~

Maybe the first is the best?

**Reply by hchiba1** [reply]
Yes, edge IDs could be supported.

And I didn’t realize the second option.
I like it because it is easy to read.
```
x: a -> b
```
